### PR TITLE
Observe auth changes in and simplify AuthProvider

### PR DESCRIPTION
**Pull request by sqs**
This is the auth parts of https://github.com/sourcegraph/cody/pull/5221. Previously, there were many ways to mutate and read auth state. Now, there is a single `authStatus` to read, and a single `authProvider.validateAndStoreCredentials` to attempt signing in.

## Test plan

e2e and CI. Also try switching user accounts and signing out, and confirm that it works.

### CODE REVIEW

**Review by dominiccooney — COMMENTED** [review]
Posting some in-progress comments to refresh the diff

File: vscode/src/services/LocalStorageProvider.ts [hunk 1]
```diff
@@ -276,10 +277,12 @@ class LocalStorage {
         return this.storage.get(key, null)
     }
 
-    public async set<T>(key: string, value: T): Promise<void> {
+    public async set<T>(key: string, value: T, fire = true): Promise<void> {
         try {
             await this.storage.update(key, value)
-            this.onChange.fire()
+            if (fire) {
```
Comment (dominiccooney) [inline vscode/src/services/LocalStorageProvider.ts:283]: This API seems prone to abuse.

The use case is legitimate, let me suggest something like:

```
localstorage.batchChangeEvent(() => {
  localStorage.set('foo', 'bar')
  localStorage.set('baz', 'quux')
})
```

You can probably think of a better name.

Within the scope of the callback (handle reentrancy) you should suppress change events and fire them when done.

This means callers can't suppress fire events unilaterally, only batch them. And their batches are limited to the same microtask...if they have async work to do, they have to prepare it first. AND it means you won't have a half-written state, interrupted by async, with someone else doing an async write and causing half written state to be observed. AND you can handle being robust to execptions etc. without forgetting to fire the event.

**Review by dominiccooney — APPROVED** [review]
Some more feedback inline.

Looks good on the whole, although it is big.

Using lodash isEqual instead of JSON serialization equality looks nice, being insensitive to property order and handling various specific types better.

**Comment by sqs** [comment]
@dominiccooney Thanks, I am focusing on getting https://github.com/sourcegraph/cody/pull/5638 merged instead for now. Then will follow up with this.

File: lib/shared/src/misc/observable.test.ts [hunk 2]
```diff
@@ -580,3 +582,91 @@ describe('concat', () => {
...
+        await done
+        expect(values).toStrictEqual<typeof values>([])
+    })
+
+    test('does not emit until other observable emits', async () => {
+        vi.useFakeTimers()
+        const source = observableOfTimedSequence(10, 'a', 20, 'b')
+        const other = observableOfTimedSequence(25, 'x')
+
+        const { values, done } = readValuesFrom(source.pipe(withLatestFrom(other)))
```
[the remark this answers is not in the document]
Comment (sqs) [reply lib/shared/src/misc/observable.test.ts:619]: The point is to be able to retrieve the values-so-far before the observable is complete.

File: vscode/src/auth/account-menu.ts [hunk 3]
```diff
@@ -41,29 +41,25 @@ enum AccountMenuOptions {
 async function openAccountMenuFirstStep(
     authStatus: AuthenticatedAuthStatus
 ): Promise<AccountMenuOptions | undefined> {
-    const isOffline = !!authStatus.isOfflineMode
-    const isDotComInstance = isDotCom(authStatus.endpoint) && !isOffline
+    const isDotComInstance = isDotCom(authStatus.endpoint)
```
[the remark this answers is not in the document]
Comment (sqs) [reply vscode/src/auth/account-menu.ts:44]: The intent of #5221 is that if you are temporarily offline, you would still be able to use local models without switching modes. Offline is an ephemeral status, not a mode.

**Comment by sqs** [comment]
closing in favor of just merging it all in https://github.com/sourcegraph/cody/pull/5221 (except those PRs that I can more cleanly separate out). the benefit of atomicity is not worth it.

**Pull request closed without merging**